STRI 1,2,4,5,6,7, semi-automated deployments

silence48 opened the pull request:
These commits are the minimum fixes for stri 1,2,4,7

silence48 committed:
Minimum fix for STRI 4. - requires exact upload fee

silence48 committed:
addresses stri 7 this validates that the xdr is valid.

silence48 committed:
this validates that the function fields are valid json or throws an err

silence48 committed:
Addresses STRI 1 + 2
Adds: SLS_TIMEOUT env var
 - Prevent multiple contracts from running at one time without having enough money
 - Adds const reserveFee = new BigNumber(SLS_TIMEOUT).dividedBy(RUN_DIVISOR).toFixed(7)
 - Hold a fee in reserver for the maximum amount of possible charges based on the
   SLS_TIMEOUT and RUN_DIVISOR env vars
 - subtracts the reserve fee from the current fee balance for the auth'ed hash.
   after execution unused fees are refunded, and cost returned.

silence48 committed:
update build system to use esbuild-loader and update lodash for lodash-es

silence48 committed:
remove residual docs github actions definition

silence48 committed:
refactor from alpha to public

silence48 committed:
update package.json and add license

silence48 committed:
update dotenv template file.

silence48 committed:
update wrangler.toml.dist template file for new bindings

silence48 committed:
remove run-script-os

silence48 committed:
Assisted automated deployment.
This commit brings automated deployment of the wrangler CLI in one step.
This is still slightly a WIP but it does work.
To deploy a turret, go to the wrangler folder and run "node deploy.mjs"
You might need to have turned on the node --experimental-vm-modules
Please use node version 16.13.0 or higher.

 On branch template
 Changes to be committed:
	modified:   ../README.md
	new file:   deploy.mjs
	new file:   deployrest.mjs
	modified:   package.json
	modified:   wrangler.toml.dist

 Changes not staged for commit:
	modified:   package-lock.json

 Untracked files:
	curlrequests.ps
	deploy/
	foralex.mjs
   tests.js
   ci.js
   sockets/

silence48 committed:
Merge pull request #1 from silence48/template

Assisted automated deployment.

silence48 commented:
(manual) automated deployment scripts have been added.   commit [7a70823 ](https://github.com/stellar/stellar-turrets/pull/2/commits/7a70823bd9d2c0a0c3205c5e1f1052b08212efb4) modifies the [wrangler.toml.dist ](wrangler/wrangler.toml.dist)... be careful; These scripts are still a work in progress.  only deploy.js is functional.  
This automatically walks the turret host through deploying their API and worker.

mootz12 commented on wrangler/src/txFunctions/run.js:
```diff
@@ -45,27 +46,45 @@ export default async ({ request, params, env }) => {
+  const reserveFee = new BigNumber(SLS_TIMEOUT).dividedBy(RUN_DIVISOR).toFixed(7)
+
   if (feeMetadata) {
     feeBalance = new BigNumber(feeMetadata.balance)
-
-    if (feeBalance.isLessThanOrEqualTo(0)) {
-      throw { status: 402, message: `Turret fees have been spent for account ${authedPublicKey}` }
+    
+    if (feeBalance.isLessThanOrEqualTo( reserveFee )) {
+      throw {
+        status: 402,
+        message: `You must have a minimum balance of ${reserveFee}.  Current FeeBalance of ${authedPublicKey} is ${feeBalance}`,
+      }
     }
+    
+    const { balance: reserve } = await txFeesStub.fetch("/", {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        minus: reserveFee,
+      }),
+    }).then(handleResponse);
```
`STRI-1`

Seems like a necessary evil. Wonder if discounting the full `MAX-FEE` is required. We might be able to get away with a smaller fee reduction, since it would still protect against the 10k concurrent request attack.

Regardless, looks good for now!

mootz12 replied:
Ah, looks like `STRI-2` is why we setup `MAX-FEE` to be the maximum amount chargeable. So disregard the suggestion to lower the reserve fee.

Just an opinion, but as a turret owner, I would be fine having users hit the negative for at most `MAX-FEE`.

silence48 replied:
> Just an opinion, but as a turret owner, I would be fine having users hit the negative for at most `MAX-FEE`.

We could make this be an option during deployment?

mootz12 replied:
I think making it an option during deployment would add an unnecessary complexity

mootz12 commented on wrangler/src/txFunctions/run.js:
```diff
@@ -110,14 +129,14 @@ export default async ({ request, params, env }) => {
       },
       body: JSON.stringify({
-        minus: cost
+        plus: settleFee
```
`STRI-1`

A comment noting that you are purposefully adding back to the users fee balance seems useful since its fairly counter intuitive

mootz12 commented on wrangler/src/txFunctions/upload.js:
```diff
@@ -42,8 +44,8 @@ export default async ({ request, env }) => {
 
-    // throws if payment fails
-    await processFeePayment(env, txFunctionFee, cost);
+    // throws if payment fails, if the fee is invalid, if the fee is too large or too small. Fixes STRI 4
+    await processFeePayment(env, txFunctionFee, cost, cost);
```
We should remove the comment tag with the `Fixes STRI-4`

That won't make much sense to someone reading the code without having read the audit

mootz12 reviewed (commented):
LG2M.

Left some comments on where we can clarify the new `Reserve Fee` implementation that fixes STRI-1 and STRI-2. Fixes for STRI-4 and STRI-7 were straight forward, nice job!

For other reviewers, Script3 has been using this version of the Turrets for the YieldBlox Testnet Beta (started ~3 days ago). So far so good. Haven't stress testing STRI-1, but general usage appears to be functioning as exepcted.

silence48 committed:
remove docs folder

silence48 committed:
Rewrite Heal Function
This commit rewrites the heal function and moves the endpoint to
/trust/heal

silence48 committed:
Update package.json, add toml

silence48 commented:
I have pushed an updated "trust" endpoint with "trust/heal" 
heal is accepts a put method and the should include the following in the request:
{sourceaccount, //the account to be healed
oldturret, // the turret to remove
newturret, //the turret to add
functionhash} // the txfunctionhash

mootz12 commented on docs/openapi/openapi.yml:
```diff
@@ -1,51 +0,0 @@
-openapi: 3.1.0
```
I don't think we should blast this.

The `stellar-turrets-docs` expects a correct `openapi.yml` file that this docs folder generates / keeps up to date

mootz12 replied:
Note - this should be blasted. Looks like docs repo will be the home for everything to avoid confusion.

mootz12 commented on wrangler/src/trust/heal.js:
```diff
@@ -0,0 +1,15 @@
+    functionhash 
+  } = await request.json()
+  
+  healresponse = await heal(sourceaccount, oldturret, newturret, functionhash)
```
unsure why the compiler let you avoid placing a `let` or `const` here.

Add one for consistency?

mootz12 commented on wrangler/src/trust/trust.js:
```diff
@@ -0,0 +1,271 @@
+        if (!sourceaccountmanagedatasigners.turret.includes(oldturret))
+            throw (`The old turret is not listed as a signer in the manage data entries`)
+        /* get the signer address to remove from the managedata entries on the source account */
+        const signertoremoveindex = sourceaccountmanagedatasigners.turret.findIndex(checkaddr);
+        const removesigner = sourceaccountmanagedatasigners.signer[signertoremoveindex]
+        function checkaddr(key) {
+            return key === oldturret;
+        }
```
const signertoremoveindex = sourceaccountmanagedatasigners.turret.findIndex((signer) => signer === oldturret);

would allow us to get rid of `checkaddr` function

mootz12 commented on wrangler/src/trust/trust.js:
```diff
@@ -0,0 +1,271 @@
+import { parse } from 'toml'
+import localtoml from '../turret/toml'
+
+/**
+ * It gets the account info for a given public key.
+ * @param pubkey - The public key of the account you want to get information about.
+ * @returns The JSON response from the Horizon API.
+ */
+export async function getAccountInfo(pubkey){
+    const accountInfo = (await fetch(`${HORIZON_URL}/accounts/${pubkey}`))
+    return accountInfo.json()
+}
```
we call this function for nearly every piece of  account info we need, likely resulting in excess Horizon calls.

Might make sense to write a `getAccount` function in `stellar-sdk-utils` that fetches and returns an accountResponse.

Said response could then be re-used throughout the majority of the utils

silence48 replied:
yes it would be smart to cache the account info

mootz12 commented on wrangler/src/trust/trust.js:
```diff
@@ -0,0 +1,271 @@
+    return accountInfo.json()
+}
+
+/**
+ * It gets all the accounts that are signed on by the given public key.
+ * @param pubkey - The public key of the signer you want to get the accounts of.
+ * @returns An array of account IDs that are signed by the given public key.
+ */
+export async function getAccountsOfSigner(pubkey) {
+    const signedOnAccounts = []
+    const getrecords = async(url) => await fetch(url)
+    const signerinfo = await (await getrecords(`${HORIZON_URL}/accounts/?signer=${pubkey}&limit=200`)).json()
+    signerinfo._embedded.records.map((record) => {
+        signedOnAccounts.push(record.account_id)
+    })
+    return signedOnAccounts
+}
```
unused

silence48 replied:
correct this is unused. there may be one other unused util function also, they were used in my create source account call which was decided to not be included, thus any unused calls can be removed.

mootz12 commented on wrangler/src/trust/trust.js:
```diff
@@ -0,0 +1,271 @@
+        const signerweight = await getSignerWeight(removesigner, sourceaccount)
+        /* check the local toml to validate the new turret is part of the trust quorem and the old turret is not.*/
+        const newtrust = await checkLocalQuorem(newturret)
+        if (!newtrust){
```
running eslint or prettier through there will fix small style typos like this automatically

if you don't have that configured just lmk and I can push a style commit through on this branch

silence48 replied:
I do have a eslint config but not on this version, I appreciate if you add one.

mootz12 reviewed (changes requested):
Looks good! The overall logic of heal looks sound. Mostly stylistic / code cleanup based feedback.

Only real feedback is this repo is still responsible for generating `openapi` files, so we shouldn't delete the docs folder

mootz12 commented on wrangler/src/trust/trust.js:
```diff
@@ -0,0 +1,271 @@
+export async function checkLocalQuorem(turret) {
+    const localquorem = localtoml()
+    try{
+        const thetoml = parse(localquorem).TSS.TURRETS
+        const quorem = []
+        for (const each in thetoml){
+            quorem.push(thetoml[each])
+        }
+        if(quorem.includes(turret))
+            return true
+        return false
```
was the TOML parsing going to be for the new TOML format we discussed?

It appears this is still the old format

silence48 replied:
Let's do the new format.

silence48 committed:
add missing const

silence48 committed:
Update the TOML parser in heal for new format

This commit updates the toml parser in the heal function to use the
following format for future turret tomls.
```toml
[[TURRETS]]
NAME="script3-mainnet"
HOME_DOMAIN="www.turret.script3.io"
PUBLIC_KEY="GABMKJM6I25XI4K7U6XWMULOUQIQ27BCTMLS6BYYSOWKTBUXVRJSXHYQ"
[[TURRETS]]
NAME="script3-mainnet"
HOME_DOMAIN="www.turret.script3.io"
PUBLIC_KEY="GABMKJM6I25XI4K7U6XWMULOUQIQ27BCTMLS6BYYSOWKTBUXVRJSXHYQ"
```

mootz12 committed:
cache control account response in heal

silence48 committed:
Merge pull request #2 from silence48/code-linting

cache control account response in heal
